### PR TITLE
Added toolchain abstraction for coverage

### PR DESCRIPTION
**Pull request by vanwinkeljan**
~~Added compiler and linker flags to generate gcov compatible coverage results with clang.~~

Added toolchain abstraction for coverage for both gcc and clang.

Note that without this support sanitycheck can not run without error when llvm toolchain is used.

### CODE REVIEW

**Review by aescolar — DISMISSED** [review]
I'd say we should be moving this into 
cmake/compiler/host-gcc/target.cmake
cmake/compiler/clang/target.cmake
under and `if ("${ARCH}" STREQUAL "posix")`

File: arch/posix/CMakeLists.txt [hunk 1]
```diff
@@ -30,6 +32,17 @@ zephyr_link_libraries_ifdef(CONFIG_COVERAGE
...
+
+zephyr_compile_options_ifdef(CONFIG_COVERAGE
+	--coverage
+	-fno-inline
+)
+zephyr_link_libraries_ifdef(CONFIG_COVERAGE
+	--coverage
+)
+endif()
+
```
Comment (SebastianBoe) [inline arch/posix/CMakeLists.txt]: We use a different mechanism to achieve portability.

See

https://github.com/zephyrproject-rtos/zephyr/commit/01592071f1e205e7a020d9dd487c04943226c287

for an example.

**Review by SebastianBoe — DISMISSED** [review]

**Comment by SebastianBoe** [comment]
> under and if ("${ARCH}" STREQUAL "posix")

If the coverage API is invoked from arch/posix/CMakeLists.txt (which it should be), then an arch check is not necessary.

**Comment by aescolar** [comment]
> If the coverage API is invoked from arch/posix/CMakeLists.txt (which it should be), then an arch check is not necessary.

True story :), I guess we just need to be careful to not mix it with the coverage for target

**Comment by SebastianBoe** [comment]
The coverage for target?

**Comment by aescolar** [comment]
> The coverage for target?

Support for `CONFIG_COVERAGE` with "real targets" was added a couple of months ago:
https://docs.zephyrproject.org/latest/guides/coverage.html#test-coverage-reports-in-embedded-devices-or-qemu

**Comment by vanwinkeljan** [comment]
@SebastianBoe @aescolar  Added toolchain abstraction for coverage for both host and target

**Comment by vanwinkeljan** [comment]
Updated PR to remove redunat linker option set in sanitycheck script

**Review by aescolar — COMMENTED** [review]
No complaints from me
The sanitycheck change is perfectly ok.
But as this is mostly cmake related, I let @SebastianBoe have the final say

File: cmake/compiler/clang/target_coverage.cmake [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+
+macro(toolchain_cc_coverage)
+
+if (CONFIG_COVERAGE)
+
+zephyr_compile_options(
```
Comment (SebastianBoe) [inline cmake/compiler/clang/target_coverage.cmake]: 2 spaces of indentation.

File: cmake/compiler/clang/target_coverage.cmake [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+
+macro(toolchain_cc_coverage)
+
+if (CONFIG_COVERAGE)
```
Comment (SebastianBoe) [inline cmake/compiler/clang/target_coverage.cmake]: This if-statement belongs in the CMakeLists.txt files.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
Requesting changes

**Review by SebastianBoe — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Comment by vanwinkeljan** [comment]
open/closing to kick CI

**Pull request merged**